util/query.py

```diff
@@ -102,6 +103,188 @@ def poolSharestoValue(shares: float, total_shares: float, base_token_liquidity:
+                totalAmountComing = 0.0
+                for delegate in user["delegates"]:
+                    totalAmountComing += float(delegate["amount"])
+
+                ## calculate total amount
+                totalAmount = totalAmountComing - totalAmountGoing
+
+                ## convert wei to OCEAN
+                totalAmount = totalAmount / 1e18
+
+                ## add to balance
+                balance += totalAmount
+
+                ## set user balance
+                if user["id"] not in veBalances:
+                    veBalances[user["id"]] = balance
+
+                veBalances[user["id"]] = (balance + veBalances[user["id"]]) / 2
+
+            ## increase offset
+            offset += chunk_size
+        n_blocks_sampled += 1
+
+    assert n_blocks_sampled > 0
+
+    print("getveBalances: done")
+
+    return veBalances
+
+
+@enforce_types
+def getAllocations(rng: BlockRange, CHAINID: int) -> dict:
+    """
+    @description
+      Return all allocations.
+
+    @return
+      allocations -- dict of [chain_id][nft_addr][user_addr]: percent
+    """
+
+    _allocations: Dict[int, Dict[str, Dict[str, float]]] = {}
+    n_blocks = rng.numBlocks()
+    n_blocks_sampled = 0
+    blocks = rng.getBlocks()
+
+    for block_i, block in enumerate(blocks):
+
+        if (block_i % 50) == 0 or (block_i == n_blocks - 1):
+            print(f"  {(block_i+1) / float(n_blocks) * 100.0:.1f}% done")
+
+        offset = 0
+        chunk_size = 1000
+        while True:
+            query = """
+          {
+            veAllocateUsers(first: %d, skip: %d, block:{number:%d}) {
+              id
+              veAllocation {
+                id
+                allocated
+                chainId
+                nftAddress
+              }
+              allocatedTotal
+            }
+          }
+          """ % (
+                chunk_size,
+                offset,
+                block,
+            )
+            result = submitQuery(query, CHAINID)
+            allocations = result["data"]["veAllocateUsers"]
+            if len(allocations) == 0:
+                # means there are no records left
+                break
+
+            for allocation in allocations:
+                user_addr = allocation["id"]
+                allocated_total = float(allocation["allocatedTotal"])
+                if user_addr not in _allocations:
+                    _allocations[user_addr] = {}
+                for ve_allocation in allocation["veAllocation"]:
+                    nft_addr = ve_allocation["nftAddress"]
+                    chain_id = ve_allocation["chainId"]
+                    allocated = float(ve_allocation["allocated"])
+                    if chain_id not in _allocations:
+                        _allocations[chain_id] = {}
+                    if nft_addr not in _allocations[chain_id]:
+                        _allocations[chain_id][nft_addr] = {}
+
+                    percentage = allocated / allocated_total
```

Comment:
This all makes sense.
Structurally, it looks correct.
And I think if we compare 2 slices of time against each other (like you did in your test), this will work.

However, what happens if:
- I allocated 2 years ago, but continue to update my voting power?

When it comes to doing the final calculation for DF reward distribution, I believe we'll have to get historical data.
Is that correct?